test/testsymboldatabase.cpp

```diff
@@ -6804,6 +6804,9 @@ class TestSymbolDatabase: public TestFixture {
         ASSERT_EQUALS("AB *", typeOf("struct AB {int a; int b;}; struct AB ab; x=&ab;", "&"));
         ASSERT_EQUALS("A::BC *", typeOf("namespace A { struct BC { int b; int c; }; }; struct A::BC abc; x=&abc;", "&"));
         ASSERT_EQUALS("A::BC *", typeOf("namespace A { struct BC { int b; int c; }; }; struct A::BC *abc; x=abc+1;", "+"));
+        ASSERT_EQUALS("signed int", typeOf("auto a(int& x, int& y) { return x + y; }", "+"));
+        ASSERT_EQUALS("signed int &", typeOf("void a(int& x, int& y) { x = y; }", "="));
+        ASSERT_EQUALS("signed int &", typeOf("auto a(int* y) { return *y; }", "*"));
```

Comment:
unfortunately I believe this will test the first `"*"`. I think it's a bit unfortunate if the valuetype is set in a variable declaration but I guess.

I am not sure but shouldn't the type be "signed int" when you dereference a "signed int *"?

Reply:
What is the best choice for cppcheck I can't say but in c++ the answer is no. For demonstration see: https://godbolt.org/z/Ms1qz6

For a logical justification, consider the statement ```*x = 4;``` if *x produces a T rather than a T& then the assignment does nothing.

It occurs to me that the rule is probably different in c where references do not exist. Does any action need to be taken to address this? I haven't considered the c use case at all in this PR as I never write c. I guess most of the changes here should be conditional on the language? At the very least the answer to this specific question seems to be different.

Reply:
hmm.. it seems logical to be consistent with the clast ast dump. And in that neither `*p` or `=` results are references.

Your test code:
```
danielm@edge:~/cppcheck$ cat 1.cpp 
auto foo(int *p)
{
    return *p;
}
danielm@edge:~/cppcheck$ clang -cc1 -ast-dump 1.cpp
TranslationUnitDecl 0x974428 <<invalid sloc>> <invalid sloc>
.....
`-FunctionDecl 0x9b3ef8 <1.cpp:1:1, line:4:1> line:1:6 foo 'int (int *)'
  |-ParmVarDecl 0x9b3dd8 <col:10, col:15> col:15 used p 'int *'
  `-CompoundStmt 0x9b4200 <line:2:1, line:4:1>
    `-ReturnStmt 0x9b41f0 <line:3:5, col:13>
      `-ImplicitCastExpr 0x9b41d8 <col:12, col:13> 'int' <LValueToRValue>
        `-UnaryOperator 0x9b4058 <col:12, col:13> 'int' lvalue prefix '*' cannot overflow
          `-ImplicitCastExpr 0x9b4040 <col:13> 'int *' <LValueToRValue>
            `-DeclRefExpr 0x9b4020 <col:13> 'int *' lvalue ParmVar 0x9b3dd8 'p' 'int *'
```

A reference variable "ref" has the type `int &`. But when assigning to the ref variable the `=` has type `int`..
```
danielm@edge:~/cppcheck$ cat 1.cpp 
void foo(int *p)
{
    int &ref = *p;
    ref = 123;
}
danielm@edge:~/cppcheck$ clang -cc1 -ast-dump 1.cpp
TranslationUnitDecl 0x1989428 <<invalid sloc>> <invalid sloc>
.....
`-FunctionDecl 0x19c8ec0 <1.cpp:2:1, line:6:1> line:2:6 foo 'void (int *)'
  |-ParmVarDecl 0x19c8df8 <col:10, col:15> col:15 used p 'int *'
  `-CompoundStmt 0x19c9180 <line:3:1, line:6:1>
    |-DeclStmt 0x19c9108 <line:4:5, col:18>
    | `-VarDecl 0x19c8ff8 <col:5, col:17> col:10 used ref 'int &' cinit
    |   `-UnaryOperator 0x19c9098 <col:16, col:17> 'int' lvalue prefix '*' cannot overflow
    |     `-ImplicitCastExpr 0x19c9080 <col:17> 'int *' <LValueToRValue>
    |       `-DeclRefExpr 0x19c9060 <col:17> 'int *' lvalue ParmVar 0x19c8df8 'p' 'int *'
    `-BinaryOperator 0x19c9160 <line:5:5, col:11> 'int' lvalue '='
      |-DeclRefExpr 0x19c9120 <col:5> 'int' lvalue Var 0x19c8ff8 'ref' 'int &'
      `-IntegerLiteral 0x19c9140 <col:11> 'int' 123
```

.. I am a bit surprised. But well right now I have the feeling it would be a good idea to be consistent.

If a reference is used or not that will not be a direct problem for `ValueFlow` etc because we really need to handle C code also so we must have this logic when assigning to non-references etc.

Reply:
ok so it sounds like we want the result of all three of these operations to be a value rather than a reference. I think that's technically wrong and may cause bugs but they'll be bugs that already exist and fixing them may cause new bugs so I'm happy to leave it.

Is that the only concern? Are there more tests you want to see?

Reply:
> Are there more tests you want to see?

I think this is it.

Reply:
clang ast clearly says that some values are "lvalue" . this info is not stored in cppcheck ast (yet).

Reply:
> hmm.. it seems logical to be consistent with the clast ast dump. And in that neither *p or = results are references.

The clang ast shows its an lvalue:

> `-DeclRefExpr 0x9b4020 <col:13> 'int *' lvalue ParmVar 0x9b3dd8 'p' 'int *'

For C++, that would be considered a reference. Instead of describing the value category in terms of C++ references, which dont exist in C, we should use the actual value categories: lvalue(`Reference::LValue`), prvalue(`Reference::None`), and xvalue(`Reference::RValue`). With the exception of xvalue(which is not possible in C), these should be the same for both C and C++ programs.

`ValueType::str()` could work more like `typeof` which strips the reference rather than like `decltype`. That should less confusing for C, which doesn't have references.

Reply:
thanks for looking... if it makes sense to add the lvalue/rvalue info I am not against that. but I do not feel that it should replace the reference info that we have now. Clang has both such reference info and also if the value is a lvalue.

The type of a variable is not lvalue/rvalue. The result from `Variable::valueType()` should not say lvalue/rvalue.

Reply:
I pushed and then reverted d73bafa which was my work in progress attempt to have operations return l/r-value references "correctly". If you decide to pursue this it might be a useful starting point. I believe it was mostly working.